File: easygraph/functions/structural_holes/evaluation.py
import math

from easygraph.utils import *


__all__ = ["effective_size", "efficiency", "constraint", "hierarchy"]


def mutual_weight(G, u, v, weight=None):
    try:
        a_uv = G[u][v].get(weight, 1)
    except KeyError:
        a_uv = 0
    try:
        a_vu = G[v][u].get(weight, 1)
    except KeyError:
        a_vu = 0
    return a_uv + a_vu


sum_nmw_rec = {}
max_nmw_rec = {}


def normalized_mutual_weight(G, u, v, norm=sum, weight=None):
    if norm == sum:
        try:
            return sum_nmw_rec[(u, v)]
        except KeyError:
            scale = norm(
                mutual_weight(G, u, w, weight) for w in set(G.all_neighbors(u))
            )
            nmw = 0 if scale == 0 else mutual_weight(G, u, v, weight) / scale
            sum_nmw_rec[(u, v)] = nmw
            return nmw
    elif norm == max:
        try:
            return max_nmw_rec[(u, v)]
        except KeyError:
            scale = norm(
                mutual_weight(G, u, w, weight) for w in set(G.all_neighbors(u))
            )
            nmw = 0 if scale == 0 else mutual_weight(G, u, v, weight) / scale
            max_nmw_rec[(u, v)] = nmw
            return nmw


def effective_size_parallel(nodes, G, weight):
    ret = []
    for node in nodes:
        neighbors_of_node = set(G.all_neighbors(node))
        if len(neighbors_of_node) == 0:
            ret.append([node, float("nan")])
            continue
        ret.append(
            [node, sum(redundancy(G, node, u, weight) for u in neighbors_of_node)]
        )
    return ret


def effective_size_borgatti_parallel(nodes, G, weight):
    ret = []
    for node in nodes:
        # Effective size is not defined for isolated nodes
        if len(G[node]) == 0:
            ret.append([node, float("nan")])
            continue
        E = G.ego_subgraph(node)
        E.remove_node(node)
        if E.size() == 0:
            ret.append([node, 1.0])
        else:
            ret.append([node, len(E) - (2 * E.size()) / len(E)])
    return ret


def redundancy(G, u, v, weight=None):
    nmw = normalized_mutual_weight
    r = sum(
        nmw(G, u, w, weight=weight) * nmw(G, v, w, norm=max, weight=weight)
        for w in set(G.all_neighbors(u))
    )
    return 1 - r


@not_implemented_for("multigraph")
@hybrid("cpp_effective_size")
def effective_size(G, nodes=None, weight=None, n_workers=None):
    """Burt's metric - Effective Size.
    Parameters
    ----------
    G : easygraph.Graph or easygraph.DiGraph
    nodes : list of nodes or None, optional (default : None)
        The nodes you want to calculate. If *None*, all nodes in `G` will be calculated.
    weight : string or None, optional (default : None)
        The key for edge weight. If *None*, `G` will be regarded as unweighted graph.
    Returns
    -------
    effective_size : dict
        The Effective Size of node in `nodes`.
    Examples
    --------
    >>> effective_size(G,
    ...                nodes=[1,2,3], # Compute the Effective Size of some nodes. The default is None for all nodes in G.
    ...                weight='weight' # The weight key of the graph. The default is None for unweighted graph.
    ...                )
    References
    ----------
    .. [1] Burt R S. Structural holes: The social structure of competition[M].
       Harvard university press, 2009.
    """
    sum_nmw_rec.clear()
    max_nmw_rec.clear()
    effective_size = {}
    if nodes is None:
        nodes = G
    # Use Borgatti's simplified formula for unweighted and undirected graphs
    if not G.is_directed() and weight is None:
        if n_workers is not None:
            import random

            from functools import partial
            from multiprocessing import Pool

            local_function = partial(
                effective_size_borgatti_parallel, G=G, weight=weight
            )
            nodes = list(nodes)
            random.shuffle(nodes)
            if len(nodes) > n_workers * 50000:
                nodes = split_len(nodes, step=50000)
            else:
                nodes = split(nodes, n_workers)
            with Pool(n_workers) as p:
                ret = p.imap(local_function, nodes)
                res = [x for i in ret for x in i]
            effective_size = dict(res)
        else:
            for v in nodes:
                # Effective size is not defined for isolated nodes
                if len(G[v]) == 0:
                    effective_size[v] = float("nan")
                    continue
                E = G.ego_subgraph(v)
                E.remove_node(v)
                if E.size() == 0:
                    effective_size[v] = 1.0
                else:
                    effective_size[v] = len(E) - (2 * E.size()) / len(E)
    else:
        if n_workers is not None:
            import random

            from functools import partial
            from multiprocessing import Pool

            local_function = partial(effective_size_parallel, G=G, weight=weight)
            nodes = list(nodes)
            random.shuffle(nodes)
            if len(nodes) > n_workers * 30000:
                nodes = split_len(nodes, step=30000)
            else:
                nodes = split(nodes, n_workers)
            with Pool(n_workers) as p:
                ret = p.imap(local_function, nodes)
                res = [x for i in ret for x in i]
            effective_size = dict(res)
        else:
            for v in nodes:
                # Effective size is not defined for isolated nodes
                if len(G[v]) == 0:
                    effective_size[v] = float("nan")
                    continue
                effective_size[v] = sum(
                    redundancy(G, v, u, weight) for u in set(G.all_neighbors(v))
                )
    return effective_size


@not_implemented_for("multigraph")
@hybrid("cpp_efficiency")
def efficiency(G, nodes=None, weight=None, n_workers=None):
    """Burt's metric - Efficiency.
    Parameters
    ----------
    G : easygraph.Graph
    nodes : list of nodes or None, optional (default : None)
        The nodes you want to calculate. If *None*, all nodes in `G` will be calculated.
    weight : string or None, optional (default : None)
        The key for edge weight. If *None*, `G` will be regarded as unweighted graph.
    Returns
    -------
    efficiency : dict
        The Efficiency of node in `nodes`.
    Examples
    --------
    >>> efficiency(G,
    ...            nodes=[1,2,3], # Compute the Efficiency of some nodes. The default is None for all nodes in G.
    ...            weight='weight' # The weight key of the graph. The default is None for unweighted graph.
    ...            )
    References
    ----------
    .. [1] Burt R S. Structural holes: The social structure of competition[M].
       Harvard university press, 2009.
    """
    e_size = effective_size(G, nodes=nodes, weight=weight, n_workers=n_workers)
    degree = G.degree(weight=weight)
    efficiency = {n: v / degree[n] for n, v in e_size.items()}
    return efficiency


def compute_constraint_of_nodes(nodes, G, weight):
    ret = []
    for node in nodes:
        neighbors_of_node = set(G.all_neighbors(node))
        if len(neighbors_of_node) == 0:
            ret.append([node, float("nan")])
            continue
        ret.append(
            [node, sum(local_constraint(G, node, u, weight) for u in neighbors_of_node)]
        )
    return ret


@not_implemented_for("multigraph")
@hybrid("cpp_constraint")
def constraint(G, nodes=None, weight=None, n_workers=None):
    """Burt's metric - Constraint.
    Parameters
    ----------
    G : easygraph.Graph
    nodes : list of nodes or None, optional (default : None)
        The nodes you want to calculate. If *None*, all nodes in `G` will be calculated.
    weight : string or None, optional (default : None)
        The key for edge weight. If *None*, `G` will be regarded as unweighted graph.
    workers : int or None, optional (default : None)
        The number of workers calculating (default: None).
        None if not using only one worker.
    Returns
    -------
    constraint : dict
        The Constraint of node in `nodes`.
    Examples
    --------
    >>> constraint(G,
    ...            nodes=[1,2,3], # Compute the Constraint of some nodes. The default is None for all nodes in G.
    ...            weight='weight', # The weight key of the graph. The default is None for unweighted graph.
    ...            n_workers=4 # Parallel computing on four workers. The default is None for serial computing.
    ...            )
    References
    ----------
    .. [1] Burt R S. Structural holes: The social structure of competition[M].
       Harvard university press, 2009.
    """
    sum_nmw_rec.clear()
    max_nmw_rec.clear()
    local_constraint_rec.clear()
    if nodes is None:
        nodes = G.nodes
    constraint = {}

    def compute_constraint_of_v(v):
        neighbors_of_v = set(G.all_neighbors(v))
        if len(G[v]) == 0:
            constraint_of_v = float("nan")
        else:
            constraint_of_v = sum(
                local_constraint(G, v, n, weight) for n in neighbors_of_v
            )
        return v, constraint_of_v

    if n_workers is not None:
        import random

        from functools import partial
        from multiprocessing import Pool

        local_function = partial(compute_constraint_of_nodes, G=G, weight=weight)
        nodes = list(nodes)
        random.shuffle(nodes)
        if len(nodes) > n_workers * 30000:
            nodes = split_len(nodes, step=30000)
        else:
            nodes = split(nodes, n_workers)
        with Pool(n_workers) as p:
            ret = p.imap(local_function, nodes)
            constraint_results = [x for i in ret for x in i]
    else:
        constraint_results = []
        for v in nodes:
            constraint_results.append(compute_constraint_of_v(v))

    constraint = dict(constraint_results)
    return constraint


local_constraint_rec = {}


def local_constraint(G, u, v, weight=None):
    try:
        return local_constraint_rec[(u, v)]
    except KeyError:
        nmw = normalized_mutual_weight
        direct = nmw(G, u, v, weight=weight)
        indirect = sum(
            nmw(G, u, w, weight=weight) * nmw(G, w, v, weight=weight)
            for w in set(G.all_neighbors(u))
        )
        result = (direct + indirect) ** 2
        local_constraint_rec[(u, v)] = result
        return result


def hierarchy_parallel(nodes, G, weight):
    ret = []
    for v in nodes:
        E = G.ego_subgraph(v)
        n = len(E) - 1
        C = 0
        c = {}
        neighbors_of_v = set(G.all_neighbors(v))
        for w in neighbors_of_v:
            C += local_constraint(G, v, w, weight)
            c[w] = local_constraint(G, v, w, weight)
        if n > 1:
            ret.append(
                [
                    v,
                    sum(
                        c[w] / C * n * math.log(c[w] / C * n) / (n * math.log(n))
                        for w in neighbors_of_v
                    ),
                ]
            )
        else:
            ret.append([v, 0])

    return ret


@not_implemented_for("multigraph")
@hybrid("cpp_hierarchy")
def hierarchy(G, nodes=None, weight=None, n_workers=None):
    """Returns the hierarchy of nodes in the graph
    Parameters
    ----------
    G : graph
    nodes :  dict, optional (default: None)
    weight : dict, optional (default: None)
    Returns
    -------
    hierarchy : dict
        the hierarchy of nodes in the graph
    Examples
    --------
    Returns the hierarchy of nodes in the graph G
    >>> hierarchy(G)
    Reference
    ---------
    https://m.book118.com/html/2019/0318/5320024122002021.shtm
    """
    sum_nmw_rec.clear()
    max_nmw_rec.clear()
    local_constraint_rec.clear()
    if nodes is None:
        nodes = G.nodes
    hierarchy = {}
    if n_workers is not None:
        import random

        from functools import partial
        from multiprocessing import Pool

        local_function = partial(hierarchy_parallel, G=G, weight=weight)
        nodes = list(nodes)
        random.shuffle(nodes)
        if len(nodes) > n_workers * 30000:
            nodes = split_len(nodes, step=30000)
        else:
            nodes = split(nodes, n_workers)
        with Pool(n_workers) as p:
            ret = p.imap(local_function, nodes)
            res = [x for i in ret for x in i]
        hierarchy = dict(res)
    else:
        for v in nodes:
            E = G.ego_subgraph(v)
            n = len(E) - 1
            C = 0
            c = {}
            neighbors_of_v = set(G.all_neighbors(v))
            for w in neighbors_of_v:
                C += local_constraint(G, v, w, weight)
                c[w] = local_constraint(G, v, w, weight)
            if n > 1:
                hierarchy[v] = sum(
                    c[w] / C * n * math.log(c[w] / C * n) / (n * math.log(n))
                    for w in neighbors_of_v
                )
            if v not in hierarchy:
                hierarchy[v] = 0
    return hierarchy

File: easygraph/classes/GraphC/Evaluation.cpp
#include "Evaluation.h"
#include "Utils.h"

struct pair_hash
{
	template<class T1, class T2>
	std::size_t operator() (const std::pair<T1, T2>& p) const
	{
		auto h1 = std::hash<T1>()(p.first);
		auto h2 = std::hash<T2>()(p.second);
		return h1 ^ h2;
	}
};

std::unordered_map<std::pair<Graph::node_t, Graph::node_t>, Graph::weight_t, pair_hash> sum_nmw_rec, max_nmw_rec, local_constraint_rec;

enum norm_t {
	sum, max
};


Graph::weight_t mutual_weight(Graph::adj_dict_factory& G, Graph::node_t u, Graph::node_t v, std::string weight) {
	Graph::weight_t a_uv = 0, a_vu = 0;
	if (G.count(u) && G[u].count(v)) {
		Graph::edge_attr_dict_factory& guv = G[u][v];
		a_uv = guv.count(weight) ? guv[weight] : 1;
	}
	if (G.count(v) && G[v].count(u)) {
		Graph::edge_attr_dict_factory& gvu = G[v][u];
		a_uv = gvu.count(weight) ? gvu[weight] : 1;
	}
	return a_uv + a_vu;
}

Graph::weight_t  normalized_mutual_weight(Graph::adj_dict_factory& G, Graph::node_t u, Graph::node_t v, std::string weight, norm_t norm = sum) {
	std::pair<Graph::node_t, Graph::node_t> edge = std::make_pair(u, v);
	auto& nmw_rec = (norm == sum) ? sum_nmw_rec : max_nmw_rec;
	if (nmw_rec.count(edge)) {
		return nmw_rec[edge];
	}
	else {
		Graph::weight_t scale = 0;
		for (auto& w : G[u]) {
			Graph::weight_t temp_weight = mutual_weight(G, u, w.first, weight);
			scale = (norm == sum) ? (scale + temp_weight) : std::max(scale, temp_weight);
		}
		Graph::weight_t nmw = scale ? (mutual_weight(G, u, v, weight) / scale) : 0;
		nmw_rec[edge] = nmw;
		return nmw;
	}
}

Graph::weight_t local_constraint(Graph::adj_dict_factory& G, Graph::node_t u, Graph::node_t v, std::string weight = "None") {
	std::pair<Graph::node_t, Graph::node_t> edge = std::make_pair(u, v);
	if (local_constraint_rec.count(edge)) {
		return local_constraint_rec[edge];
	}
	else {
		Graph::weight_t direct = normalized_mutual_weight(G, u, v, weight);
		Graph::weight_t indirect = 0;
		for (auto& w : G[u]) {
			indirect += normalized_mutual_weight(G, u, w.first, weight) * normalized_mutual_weight(G, w.first, v, weight);
		}
		Graph::weight_t result = pow((direct + indirect), 2);
		local_constraint_rec[edge] = result;
		return result;
	}
}

std::pair<Graph::node_t, Graph::weight_t> compute_constraint_of_v(Graph::adj_dict_factory& G, Graph::node_t v, std::string weight) {
	Graph::weight_t constraint_of_v = 0;
	if (G[v].size() == 0) {
		constraint_of_v = Py_NAN;
	}
	else {
		for (const auto& n : G[v]) {
			constraint_of_v += local_constraint(G, v, n.first, weight);
		}
	}
	return std::make_pair(v, constraint_of_v);
}

py::object constraint(py::object G, py::object nodes, py::object weight, py::object n_workers) {
	std::string weight_key = weight_to_string(weight);
	sum_nmw_rec.clear();
	max_nmw_rec.clear();
	local_constraint_rec.clear();
	if (nodes == py::object()) {
		nodes = G.attr("nodes");
	}
	py::list nodes_list = py::list(nodes);
	py::list constraint_results = py::list();
	Graph& G_ = py::extract<Graph&>(G);
	for (int i = 0;i < py::len(nodes_list);i++) {
		py::object v = nodes_list[i];
		Graph::node_t v_id = py::extract<Graph::node_t>(G_.node_to_id[v]);
		std::pair<Graph::node_t, Graph::weight_t> constraint_pair = compute_constraint_of_v(G_.adj, v_id, weight_key);
		py::tuple constraint_of_v = py::make_tuple(G_.id_to_node[constraint_pair.first], constraint_pair.second);
		constraint_results.append(constraint_of_v);
	}
	py::dict constraint = py::dict(constraint_results);
	return constraint;
}

Graph::weight_t redundancy(Graph::adj_dict_factory& G, Graph::node_t u, Graph::node_t v, std::string weight = "None") {
	Graph::weight_t r = 0;
	for (const auto& neighbor_info : G[u]) {
		Graph::node_t w = neighbor_info.first;
		r += normalized_mutual_weight(G, u, w, weight) * normalized_mutual_weight(G, v, w, weight, max);
	}
	return 1 - r;
}

py::object effective_size(py::object G, py::object nodes, py::object weight, py::object n_workers) {
	Graph& G_ = py::extract<Graph&>(G);
	sum_nmw_rec.clear();
	max_nmw_rec.clear();
	py::dict effective_size = py::dict();
	if (nodes == py::object()) {
		nodes = G;
	}
	nodes = py::list(nodes);
	if (!G.attr("is_directed")() && weight == py::object()) {
		for (int i = 0;i < py::len(nodes);i++) {
			py::object v = nodes[i];
			if (py::len(G[v]) == 0) {
				effective_size[v] = py::object(Py_NAN);
				continue;
			}
			py::object E = G.attr("ego_subgraph")(v);
			if (py::len(E) > 1) {
				Graph::weight_t size = py::extract<Graph::weight_t>(E.attr("size")());
				effective_size[v] = py::len(E) - 1 - (2.0 * size) / (py::len(E) - 1);
			}
			else {
				effective_size[v] = 0;
			}
		}
	}
	else {
		std::string weight_key = weight_to_string(weight);
		for (int i = 0;i < py::len(nodes);i++) {
			py::object v = nodes[i];
			if (py::len(G[v]) == 0) {
				effective_size[v] = py::object(Py_NAN);
				continue;
			}
			Graph::weight_t redundancy_sum = 0;
			Graph::node_t v_id = py::extract<Graph::node_t>(G_.node_to_id[v]);
			for (const auto& neighbor_info : G_.adj[v_id]) {
				Graph::node_t u_id = neighbor_info.first;
				redundancy_sum += redundancy(G_.adj, v_id, u_id, weight_key);
			}
			effective_size[v] = redundancy_sum;
		}
	}
	return effective_size;
}

py::object hierarchy(py::object G, py::object nodes, py::object weight, py::object n_workers) {
	sum_nmw_rec.clear();
	max_nmw_rec.clear();
	local_constraint_rec.clear();
	std::string weight_key = weight_to_string(weight);
	if (nodes == py::object()) {
		nodes = G.attr("nodes");
	}
	py::list nodes_list = py::list(nodes);

	Graph& G_ = py::extract<Graph&>(G);
	py::dict hierarchy = py::dict();

	for (int i = 0;i < py::len(nodes_list);i++) {
		py::object v = nodes_list[i];
		py::object E = G.attr("ego_subgraph")(v);

		int n = py::len(E) - 1;

		Graph::weight_t C = 0;
		std::map<Graph::node_t, Graph::weight_t> c;
		py::list neighbors_of_v = py::list(G.attr("neighbors")(v));

		for (int j = 0;j < py::len(neighbors_of_v);j++) {
			py::object w = neighbors_of_v[j];
			Graph::node_t v_id = py::extract<Graph::node_t>(G_.node_to_id[v]);
			Graph::node_t w_id = py::extract<Graph::node_t>(G_.node_to_id[w]);
			C += local_constraint(G_.adj, v_id, w_id, weight_key);
			c[w_id] = local_constraint(G_.adj, v_id, w_id, weight_key);
		}
		if (n > 1) {
			Graph::weight_t hierarchy_sum = 0;
			for (int k = 0;k < py::len(neighbors_of_v);k++) {
				py::object w = neighbors_of_v[k];
				Graph::node_t w_id = py::extract<Graph::node_t>(G_.node_to_id[w]);
				hierarchy_sum += c[w_id] / C * n * log(c[w_id] / C * n) / (n * log(n));
			}
			hierarchy[v] = hierarchy_sum;
		}
		if (!hierarchy.has_key(v)) {
			hierarchy[v] = 0;
		}
	}
	return hierarchy;
}